### DISCUSSION TITLE
Loco Positioning: code specific documentation?

### DISCUSSION THREAD

**Discussion by karlscholz** [bitcraze/discussions#186] [Q&A] [answered]
Does anyone know where I can find a specific code documentation for the loco system?
I found these:

    https://www.bitcraze.io/documentation/system/positioning/loco-positioning-system/
    https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/
    https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/

But none of them are very close to the code. Is there another documentation that helps me understand better which function calls what and how it is implemented in the main system loop or do you have any tips for me on how and where to start with reverse engineering the code?

**Comment by krichardsson** [comment, marked as answer]
I think you found most of the documentation already. There is some more in the [crazyflie-firmware repo docs](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/).

Unfortunately there is no documentation describing the implementation but I'll give you some pointers:
* The [Deck driver for the Loco deck](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/deck/drivers/src/locodeck.c) is the starting point of the functionality.
* The drivers for the DW1000 chip is implemented in the [libdw1000 repo](https://github.com/bitcraze/libdw1000)
* Implementations for [TWR](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/deck/drivers/src/lpsTwrTag.c), [TDoA2](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/deck/drivers/src/lpsTdoa2Tag.c) and [TDoA3](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/deck/drivers/src/lpsTdoa3Tag.c) can be found in the same directory 
* The TDoA flavors above only implements the protocol while the distance part is handled by the [TDoA engine](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/utils/src/tdoa/tdoaEngine.c)
* Measurements are fused with other sensor data in the [Kalman estimator](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/estimator_kalman.c), where the actual computations are done in the [kalman core](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/kalman_core/kalman_core.c)
* Implementation of the measurement models are in the same directory, both for [TWR](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/kalman_core/mm_distance.c) and [TDoA](https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/kalman_core/mm_tdoa.c)

I hope this gives you an idea of where to start

**Reply by karlscholz** [reply]
Thanks for the hint with the kalman core :)
